lib/compat/wordpress-5.9/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -144,7 +144,11 @@ public static function get_theme_data( $deprecated = array() ) {
 		if ( null === self::$theme ) {
 			$theme_json_data = self::read_json_file( self::get_file_path_from_theme( 'theme.json' ) );
 			$theme_json_data = self::translate( $theme_json_data, wp_get_theme()->get( 'TextDomain' ) );
-			self::$theme     = new WP_Theme_JSON_Gutenberg( $theme_json_data );
+
```

Comment:
Is this PR landing for 5.9? If it's not, we'll need to figure out a different way to do this, as code in `lib/compat/wordpress-x.x` is meant to be exactly the same as the code that is backported to that WordPress version.

Reply:
Makes sense... However, I don't understand how this is supposed to work exactly. We added this because as per your comment in https://github.com/WordPress/gutenberg/pull/36394#pullrequestreview-814518096, we didn't want a filter there - which led to commit https://github.com/WordPress/gutenberg/pull/36394/commits/db3c8a158ecac0b739c95cf961d7b9fb70a7d706
If we don't have a filter or another method of hooking in there, then how can this be in a separate file? 🤔

Reply:
We need to do something for 6.0, keep an eye on the conversation at https://github.com/WordPress/gutenberg/pull/36909#issuecomment-990913068

Reply:
@oandregal Since I see #36909 was closed, do we have any ideas or plans on how to accomplish things like this? 🤔

Reply:
Just an update/note-to-self here.

I _think_ I found a way to hook in the global-styles, but only for the editor side, and it's ugly, feels too hacky.
It would involve basically overriding the `WP_REST_Server` class using the `wp_rest_server_class` filter. In our custom class we could then override the `register_route` method, and conditionally override the callbacks based on the endpoints etc.

```php
add_filter( 'wp_rest_server_class', function( $class_name ) {
	return 'Custom_WP_REST_Server';
});

class Custom_WP_REST_Server extends WP_REST_Server {
	// Override the `register_route` method here, adding lots of conditions
	// so we can add custom callbacks etc.
}
```

It would be preferable to have a filter of some sort, but in the meantime, this hack can work. It's suboptimal and sets a bad precedence, but I can't find a better alternative at this point.
What do you think @oandregal?

Reply:
What do you think of doing this instead?

- Rename the existing `WP_Theme_JSON_Resolver_Gutenberg`  to `WP_Theme_JSON_Resolver_5_9`.
- Create a class named `WP_Theme_JSON_Resolver_Gutenberg` that extends `WP_Theme_JSON_Resolver_5_9` in `lib/compat/6.0`.
- Change to `protected` everything that's `private` in `WP_Theme_JSON_Resolver_5_9` (methods and constants). This would need to be backported to 5.9.1. See https://github.com/WordPress/gutenberg/pull/38625
- Override `get_theme_data` with the things you need in `WP_Theme_JSON_Resolver_Gutenberg`.

Reply:
If we back port to `5.9.1` the changes in your PR @oandregal , that would still require to rename here to `WP_Theme_JSON_Resolver_5_9` and not check if class `WP_Theme_JSON_Resolver` exists because sites might be running `5.9.0`. Is this correct?

Reply:
Yeah, we need to do the other steps in this PR.

Reply:
I've pushed some changes here to be able to test this together with https://github.com/WordPress/gutenberg/pull/38625 

@aristath would you help me understand how to best test this? I've added the following to the empty theme's `theme.json`:

```json
"fontFamilies": [
				{
					"fontFamily": "\"Source Serif Pro\", serif",
					"name": "Source Serif Pro",
					"slug": "source-serif-pro",
					"fontFace": [
						{
							"fontFamily": "Source Serif Pro",
							"fontWeight": "200 900",
							"fontStyle": "normal",
							"fontStretch": "normal",
							"src": [ "file:./assets/fonts/SourceSerif4Variable-Roman.ttf.woff2" ]
						},
						{
							"fontFamily": "Source Serif Pro",
							"fontWeight": "200 900",
							"fontStyle": "italic",
							"fontStretch": "normal",
							"src": [ "file:./assets/fonts/SourceSerif4Variable-Italic.ttf.woff2" ]
						}
					]
				}
			]
```

And then added a list block. I understand I should be able to see the font family control with these fonts, but it's not available in the typography options.

Reply:
Thank you @oandregal!
I just tested registering webfonts from `theme.json` and that works fine. The webfonts appear in the font-family picker in the site-editor, and styles get enqueued properly both on the editor and the frontend.

What still doesn't work is webfonts registered from a plugin. Example:
```php
add_action( 'after_setup_theme', function() {
	if ( ! function_exists( 'wp_register_webfonts' ) ) {
		return;
	}
	wp_register_webfonts(
		array(
			array(
				'font-family'  => 'Random Font Name',
				'font-weight'  => '200 900',
				'font-style'   => 'normal',
				'font-stretch' => 'normal',
				'src'          => array( 'file:./assets/fonts/SourceSerif4Variable-Roman.ttf.woff2' ),
			),
			array(
				'font-family'  => 'Random Font Name',
				'font-weight'  => '200 900',
				'font-style'   => 'italic',
				'font-stretch' => 'normal',
				'src'          => array( 'file:./assets/fonts/SourceSerif4Variable-Italic.ttf.woff2' ),
			),
		)
	);
} );
```

Webfonts registered via the `wp_register_webfonts` function _should_ show up in the font-family selector, so in this case, `Random Font Name` should show up.

That is the part I'm trying to figure out at the moment 👍